packages/gatsby-react-router-scroll/src/StateStorage.js

```diff
@@ -1,16 +1,41 @@
 const STATE_KEY_PREFIX = `@@scroll|`
+const GATSBY_ROUTER_SCROLL_STATE = `___GATSBY_REACT_ROUTER_SCROLL`
 
 export default class SessionStorage {
   read(location, key) {
     const stateKey = this.getStateKey(location, key)
-    const value = sessionStorage.getItem(stateKey)
-    return JSON.parse(value)
+
+    try {
+      const value = sessionStorage.getItem(stateKey)
+      return JSON.parse(value)
+    } catch (e) {
+      console.warn(`[gatsby-react-router-scroll] Unable to access sessionStorage; sessionStorage is not available.`)
+
+      if (window && window[GATSBY_ROUTER_SCROLL_STATE] && window[GATSBY_ROUTER_SCROLL_STATE][stateKey]) {
+        return window[GATSBY_ROUTER_SCROLL_STATE][stateKey]
+      }
+
+      return ({})
+    }
   }
 
   save(location, key, value) {
     const stateKey = this.getStateKey(location, key)
     const storedValue = JSON.stringify(value)
-    sessionStorage.setItem(stateKey, storedValue)
+
+    try {
+      sessionStorage.setItem(stateKey, storedValue)
+    } catch (e) {
+      if (window && window[GATSBY_ROUTER_SCROLL_STATE]) {
```

Comment:
Same goes here too. Did the saving of the scroll position here, rather than having the check in _try_ block because browser like Safari would throw error immediately, even when just trying to check using `window.sessionStorage`

Reply:
Yeah, this is totally reasonable way of doing things. It seems it's quite rare for this error to show up as there's been lots of Gatsby sites built without me hearing about this until now :-)

So even if a try/catch is perhaps a bit slower, it's a very rare code path.